feat(subcontracting): Added provision to create multiple Subcontracting Orders against a single PO

mihir-kandoi opened the pull request:
PR to add feature requested at #43986

This feature allows multiple Subcontracting Orders to be created for a single subcontracted Purchase Order

mihir-kandoi committed:
feat(subcontracting): Added provision to have multiple Subcontracting Orders against a single PO

mihir-kandoi committed:
fix(new_sc_flow): fixed conversion factor calculations

mihir-kandoi committed:
fix(new_sc_flow): removed unneccessary indentation changes in purchase_order_item.json

rohitwaghchaure commented on erpnext/buying/doctype/purchase_order_item/purchase_order_item.json:
```diff
@@ -929,13 +930,20 @@
+   "fieldname": "sco_qty",
+   "fieldtype": "Float",
+   "label": "Subcontracted Quantity",
+   "read_only": 1
```
Enable No Copy property

rohitwaghchaure commented on erpnext/controllers/subcontracting_controller.py:
```diff
@@ -103,6 +102,13 @@ def validate_items(self):
 						_("Row {0}: Item {1} must be a subcontracted item.").format(item.idx, item.item_name)
 					)
 
+				po_item = frappe.get_doc("Purchase Order Item", item.purchase_order_item)
```
Use of get_doc in for loop is expensive, better use single db query to get required data for all items

rohitwaghchaure commented on erpnext/subcontracting/doctype/subcontracting_order/subcontracting_order.js:
```diff
@@ -5,10 +5,30 @@ frappe.provide("erpnext.buying");
+		let row = locals[cdt][cdn];
+		frm.toggle_enable(["service_items"], true);
+		frm.get_field("service_items").grid.grid_rows[row.idx - 1].remove();
+		conversion_factor.splice(row.idx - 1, 1);
```
Logic of this code is very complex, write in simple way

mihir-kandoi committed:
fix(new_sc_flow): Made requested changes

1. Removed conversion factor logic from purchase_order.js and made it into a field in SCO Item instead
2. Replaced expensive frappe.get_doc operation with query builder in subcontracting_controller.py
3. Fixed a bug where items from a list were being removed while iterating over the same list causing unexpected behaviour
4. Added "No Copy" to sco_qty field in PO Item
5. Added test cases for the enitre new_sc_flow

rohitwaghchaure commented on erpnext/controllers/subcontracting_controller.py:
```diff
@@ -89,6 +90,13 @@ def set_items_conversion_factor(self):
+					 .select(table.name, table.qty, table.sco_qty, table.fg_item_qty)
+					 .where(table.parent == self.purchase_order)
+		)
+		po_items = query.run(as_dict=True)
```
- fg_item_qty is not using so no need to select

- Make a dict po_wise_qty = {"purchase_order_item": "qty - sco_qty"} and use it instead of the below code

```
po_item = next(po_item for po_item in po_items if po_item.name == item.purchase_order_item)
if self.doctype not in "Subcontracting Receipt" and item.qty > ((po_item.qty - po_item.sco_qty) / item.sc_conversion_factor):
```

use

```
sco_qty = po_wise_qty.get(item.purchase_order_item)
if self.doctype not in "Subcontracting Receipt" and item.qty > (flt(sco_qty) / item.sc_conversion_factor)
```

mihir-kandoi committed:
fix: Made requested changes in subcontracting_controller.py

rohitwaghchaure commented on erpnext/buying/doctype/purchase_order/purchase_order.py:
```diff
@@ -868,30 +868,34 @@ def make_inter_company_sales_order(source_name, target_doc=None):
-				indicator="green",
-				alert=True,
-			)
+	is_sco_full = all(po_item.qty == po_item.sco_qty for po_item in frappe.get_doc("Purchase Order", source_name).items)
```
Only qty and sco_qty is required here from the purchase order item table then why get_doc used, maybe get the list using qb where qty != sco_qty

rohitwaghchaure commented on erpnext/controllers/subcontracting_controller.py:
```diff
@@ -89,6 +90,14 @@ def set_items_conversion_factor(self):
 				item.conversion_factor = 1
 
 	def validate_items(self):
+		table = DocType("Purchase Order Item")
+		query = (
+			frappe.qb.from_(table)
+					 .select(table.name, table.qty, table.sco_qty)
+					 .where(table.parent == self.purchase_order)
+		)
+		po_wise_qty = {item.name: item.qty - item.sco_qty for item in query.run(as_dict=True)}
+
```
Write in a separate function like get_pending_sco_qty

mihir-kandoi committed:
refactor(new_sc_flow): Made requested changes by reviewer (Rohit)

1. Replaced get_doc in purchase_order.py with Frappe QB
2. Moved logic of getting pending SCO quantity to seperate function in subcontracting_controller.py

mihir-kandoi committed:
Made requested changes

1. Replaced get_doc in purchase_order.py with Frappe QB
2. Moved logic of getting pending SCO quantity to seperate function in subcontracting_controller.py

mihir-kandoi committed:
Merge branch 'develop' into new_sc_flow

mihir-kandoi committed:
refactor(new_sc_flow): Remove unused variable from subcontracting_order.py

mihir-kandoi committed:
refactor(new_sc_flow): Made changes for PR

1. Added translation requirement when message thrown in purchase_order.py
2. Moved call to get_pending_sco_qty function to ensure it doesn't get called on Subcontracting Receipt functions in subcontracting_controller.py

mihir-kandoi committed:
Merge branch 'new_sc_flow' of https://github.com/frappe/erpnext into new_sc_flow

mihir-kandoi committed:
fix(new_sc_flow): Changes requested by pre-commit and fixed tests

mihir-kandoi committed:
chore(new_sc_flow): Made changes suggested by pre-commit

mihir-kandoi committed:
chore(new_sc_flow): Fix linter issue

mihir-kandoi committed:
chore(new_sc_flow): Fix variable name error
